lib/nostrum/shard/session/compression/zstd.ex

```diff
@@ -0,0 +1,61 @@
+defmodule Nostrum.Shard.Session.Compression.Zstd do
+  @moduledoc """
+  Implementation of compression methods for the `zstd` compression algorithm.
+
+  > ### Internal module {: .info}
+  >
+  > This module is intended for exclusive usage inside of nostrum, and is
+  > documented for completeness and people curious to look behind the covers.
+  """
+
+  @behaviour Nostrum.Shard.Session.Compression
+
+  def check_available! do
+    if not Code.ensure_loaded?(:ezstd) do
+      zstd_missing()
+    end
+  end
+
+  @spec zstd_missing() :: no_return
+  defp zstd_missing do
+    raise ArgumentError, """
+    Cannot use the :zstd gateway compression option without optional dependency :ezstd.
+
+    See https://kraigie.github.io/nostrum/gateway_compression.html for more information.
+    """
+  end
+
+  if Code.ensure_loaded?(:ezstd) do
+    @zstd_buffer_size 16_000
+
+    @spec create_context() :: reference() | {:error, any()}
+    def create_context do
+      :ezstd.create_decompression_context(@zstd_buffer_size)
+    end
+
+    @spec inflate(reference(), iodata()) :: iolist() | {:error, any()}
+    def inflate(ctx, frame) do
+      :ezstd.decompress_streaming(ctx, frame)
+    end
+
+    @spec reset_context(reference()) :: reference()
+    def reset_context(_ctx) do
+      create_context()
+    end
```

Comment:
Contexts are cleaned up by GC once the reference is no longer being referenced.

The usage of this function means that the Shard session state machine updates it's state with this new context and the old one should have no references (other than some hopefully short-lasting latent references sent to users as part of the ws state).

Once silviucpp/ezstd#16 is merged we can instead reset the context instead of having to create a new one.